File: docs/validation/python/app/scenarios/av04_missing_group.py
from pathlib import Path
from machine_config import MachineConfigReader

"""
AV-04: Missing required HDF5 group (Machine/)
ID:          AV-04
Title:       Reader returns typed error when required group is absent
Category:    adapter-versioning / error-contract
Layer:       reader (v1.0 adapter)
Precondition: docs/validation/fixtures/missing_machine_group.h5
Action:      Attempt to read/parse
Expected:    Typed error (not panic) indicating the missing group.
             Error message must name the missing path.
             No partial MachineConfig returned.
Rationale:   Corrupt or hand-edited files in the field will be missing groups.
"""

def run(fixtures_dir: str, real_dir: str) -> tuple[bool, str]:
    fixture = Path(fixtures_dir).parent / "docs" / "validation" / "fixtures" / "missing_machine_group.h5"
    try:
        MachineConfigReader(fixture).parse()
        return False, "no error raised for missing Machine/ group"
    except Exception as e:
        return True, f"{type(e).__name__} raised for missing Machine/ group: {e}"

File: docs/validation/cpp/app/src/scenarios/av04_missing_group.cpp
// AV-04: Reader returns an error when required group (Machine/) is absent
#include "av04_missing_group.hpp"
#include "machine_config/machine_config.hpp"

namespace av04 {

using namespace machine_config;

scenarios::Result run(const std::filesystem::path& fixturesDir, const std::filesystem::path&) {
    auto fixture = scenarios::avFixture(fixturesDir, "missing_machine_group.h5");
    try {
        MachineConfigReader{fixture}.parse();
        return {false, "no error raised for missing Machine/ group"};
    } catch (const std::exception& e) {
        return {true, std::string("error raised for missing Machine/ group: ") + e.what()};
    }
}

} // namespace av04
